processor/transformprocessor/internal/metrics/func_scale.go

```diff
@@ -0,0 +1,125 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package metrics // import "github.com/open-telemetry/opentelemetry-collector-contrib/processor/transformprocessor/internal/metrics"
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"go.opentelemetry.io/collector/pdata/pmetric"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl/contexts/ottlmetric"
+)
+
+type ScaleArguments struct {
+	Multiplier float64
+	Unit       ottl.Optional[ottl.StringGetter[ottlmetric.TransformContext]]
+}
+
+func newScaleMetricFactory() ottl.Factory[ottlmetric.TransformContext] {
+	return ottl.NewFactory("scale_metric", &ScaleArguments{}, createScaleFunction)
+}
+
+func createScaleFunction(_ ottl.FunctionContext, oArgs ottl.Arguments) (ottl.ExprFunc[ottlmetric.TransformContext], error) {
+	args, ok := oArgs.(*ScaleArguments)
+
+	if !ok {
+		return nil, fmt.Errorf("ScaleFactory args must be of type *ScaleArguments[K]")
+	}
+
+	return Scale(*args)
+}
+
+func Scale(args ScaleArguments) (ottl.ExprFunc[ottlmetric.TransformContext], error) {
+	return func(ctx context.Context, tCtx ottlmetric.TransformContext) (any, error) {
+		metric := tCtx.GetMetric()
+
+		switch metric.Type() {
+		case pmetric.MetricTypeGauge:
+			scaleMetric(metric.Gauge().DataPoints(), args.Multiplier)
+		case pmetric.MetricTypeHistogram:
+			scaleHistogram(metric.Histogram().DataPoints(), args.Multiplier)
+		case pmetric.MetricTypeSummary:
+			scaleSummarySlice(metric.Summary().DataPoints(), args.Multiplier)
+		case pmetric.MetricTypeSum:
+			scaleMetric(metric.Sum().DataPoints(), args.Multiplier)
+		case pmetric.MetricTypeExponentialHistogram:
+			return nil, errors.New("exponential histograms are not supported by the 'scale_metric' function")
+		default:
+			return nil, fmt.Errorf("unsupported metric type: '%v'", metric.Type())
+		}
+
+		if !args.Unit.IsEmpty() {
+			unit, err := args.Unit.Get().Get(ctx, tCtx)
+			if err != nil {
+				return nil, fmt.Errorf("could not get unit from ScaleArguments: %w", err)
+			}
+			metric.SetUnit(unit)
+		}
```

Comment:
We should do this check first - if a unit is specified and we can't retrieve it we shouldnt scale the metric.  Actually setting the unit can happen after the switch statement to ensure we only update the metric if we pass the error checks in the switch statement.

Reply:
good suggestion, I have adapted the logic accordingly